src/pubsub/ua_pubsub_writer.c

```diff
@@ -697,9 +816,18 @@ UA_Server_addDataSetField(UA_Server *server, const UA_NodeId publishedDataSet,
     else
         TAILQ_INSERT_HEAD(&currentDataSet->fields, newField, listEntry);
 
-    if(newField->config.field.variable.promotedField)
-        currentDataSet->promotedFieldsCount++;
-    currentDataSet->fieldSize++;
+    if (newField->config.dataSetFieldType == UA_PUBSUB_DATASETFIELD_VARIABLE){
+        if(newField->config.field.variable.promotedField)
+            currentDataSet->promotedFieldsCount++;
+        currentDataSet->fieldSize++;
+    } else if(newField->config.dataSetFieldType == UA_PUBSUB_DATASETFIELD_EVENT){
+        if(newField->config.field.events.promotedField)
+            currentDataSet->promotedFieldsCount++;
+        currentDataSet->fieldSize++;
```

Comment:
Is it necessary to call addDataSetField after creating an Event-PDS?

Reply:
Currently yes, it sets the correct PDS fieldSize, the promotedFields and die FieldMetaData. It wouldn't be too difficult to rewrite it though.